# REVIESED COPY OF EDA

The goal of this project is to clean, and explore the data from a csv file on my running and hopefully extract some trends

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt  #incase i wanna graph
import seaborn as sns            # more graphing
import plotly.express as px      #ngl dont even remember what this does

#read over these after and get rid of what's not needed


In [2]:
df= pd.read_csv('activities2.csv')

df

Activity ID              Activity Date  \
0      2081785707   Nov 17, 2018, 5:04:51 PM   
1      2081785716  Dec 27, 2018, 10:36:08 PM   
2      2081785724    Nov 7, 2018, 9:02:38 PM   
3      2081785725   Dec 18, 2018, 9:45:51 PM   
4      2081785728    Jan 6, 2019, 8:19:24 PM   
...           ...                        ...   
1130   8775292336   Mar 25, 2023, 2:26:39 PM   
1131   8783182130   Mar 26, 2023, 8:30:40 PM   
1132   8787353144   Mar 27, 2023, 3:33:02 PM   
1133   8788572014  Mar 27, 2023, 10:17:55 PM   
1134   8794388032   Mar 28, 2023, 9:49:46 PM   

                                   Activity Name Activity Type  \
0                                 Lunch Activity       Workout   
1                             Afternoon Activity       Workout   
2                             Afternoon Activity       Workout   
3                             Afternoon Activity       Workout   
4                             Afternoon Activity       Workout   
...                                          ...           ...   
1130  Album: Sticky fingers- Land of pleasure \n          Ride   
1131                              Takin it easy            Run   
1132                                     Cyclone    Elliptical   
1133                     Good to see tha homies            Run   
1134             Zach gate has been proven wrong           Run   

                                   Activity Description  Elapsed Time  \
0                                                   NaN          1310   
1                                                   NaN           663   
2                                                   NaN          4658   
3                                                   NaN          2924   
4                                                   NaN          2305   
...                                                 ...           ...   
1130             15 wu\n10-8-6 w 2\n15 cd\n\n90-100 rpm          3526   
1131  Didn’t go as far, hoping the calf feels a bit ...          2259   
1132                  15 wu\n8 tempo\n3\n8x2 w 1\n15 cd          3790   
1133                                                NaN          1211   
1134                                                NaN          2336   

      Distance  Max Heart Rate  Relative Effort  Commute  ...  \
0         2.81             NaN              NaN    False  ...   
1         2.14             NaN              NaN    False  ...   
2        13.72             NaN              NaN    False  ...   
3         8.82             NaN              NaN    False  ...   
4         7.85             NaN              NaN    False  ...   
...        ...             ...              ...      ...  ...   
1130      0.00           154.0             26.0    False  ...   
1131      7.60           174.0             56.0    False  ...   
1132      0.00           167.0             39.0    False  ...   
1133      3.62           161.0             16.0    False  ...   
1134      8.12           184.0             62.0    False  ...   

     <span class="translation_missing" title="translation missing: en-US.lib.export.portability_exporter.activities.horton_values.total_grit">Total Grit</span>  \
0                                                   NaN                                                                                                           
1                                                   NaN                                                                                                           
2                                                   NaN                                                                                                           
3                                                   NaN                                                                                                           
4                                                   NaN                                                                                                           


In [3]:
df.shape

(1135, 87)

Off the bat we can see we have 87 columns and just over 1,100 rows. Note, this is using a new data set taken a few weeks after the first data set was downloaded, and has more rows. 

Some of these columns are not needed so we will go through and try to get rid of the redudant ones.

In [4]:
df.columns

Index(['Activity ID', 'Activity Date', 'Activity Name', 'Activity Type',
       'Activity Description', 'Elapsed Time', 'Distance', 'Max Heart Rate',
       'Relative Effort', 'Commute', 'Activity Private Note', 'Activity Gear',
       'Filename', 'Athlete Weight', 'Bike Weight', 'Elapsed Time.1',
       'Moving Time', 'Distance.1', 'Max Speed', 'Average Speed',
       'Elevation Gain', 'Elevation Loss', 'Elevation Low', 'Elevation High',
       'Max Grade', 'Average Grade', 'Average Positive Grade',
       'Average Negative Grade', 'Max Cadence', 'Average Cadence',
       'Max Heart Rate.1', 'Average Heart Rate', 'Max Watts', 'Average Watts',
       'Calories', 'Max Temperature', 'Average Temperature',
       'Relative Effort.1', 'Total Work', 'Number of Runs', 'Uphill Time',
       'Downhill Time', 'Other Time', 'Perceived Exertion',
       '<span class="translation_missing" title="translation missing: en-US.lib.export.portability_exporter.activities.horton_values.type">Type</span>',

We can see that some of the columns have a certain error message at the begining and end we will see if we can remove these parts of the column names so that they are readable and we can deem if they are useful or not.

In [5]:
#using df info to find the affected columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1135 entries, 0 to 1134
Data columns (total 87 columns):
 #   Column                                                                                                                                                                                          Non-Null Count  Dtype  
---  ------                                                                                                                                                                                          --------------  -----  
 0   Activity ID                                                                                                                                                                                     1135 non-null   int64  
 1   Activity Date                                                                                                                                                                                   1135 non-null   object 
 2   Acti

By checking the dataframe info we can have a better view of the columns and their number so we can find a range of afffected columns

From this we can see that columns 44-45 as well as 76-85 are all affected.

# Renaming

In [6]:
df.columns[44:46]

Index(['<span class="translation_missing" title="translation missing: en-US.lib.export.portability_exporter.activities.horton_values.type">Type</span>', '<span class="translation_missing" title="translation missing: en-US.lib.export.portability_exporter.activities.horton_values.start_time">Start Time</span>'], dtype='object')

In [7]:
# using .rename to rename the columns 
# demo--> df.rename(columns=dict(zip(df.columns[old col range], new_col_names)))
#source: https://stackoverflow.com/questions/47343838/how-to-change-column-names-in-pandas-dataframe-using-a-list-of-names


new_names=['Type', 'Start Time']
df.rename(columns=dict(zip(df.columns[44:46], new_names)), inplace=True)

In [8]:
#checkgin that the names changed
df.columns[44:46]

Index(['Type', 'Start Time'], dtype='object')

NICE! It appears to have worked. Lets try the rest.

In [9]:
#Rechecking column names that still need changing
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1135 entries, 0 to 1134
Data columns (total 87 columns):
 #   Column                                                                                                                                                                                          Non-Null Count  Dtype  
---  ------                                                                                                                                                                                          --------------  -----  
 0   Activity ID                                                                                                                                                                                     1135 non-null   int64  
 1   Activity Date                                                                                                                                                                                   1135 non-null   object 
 2   Acti

The remaining columns to fix are 76-85.

In [10]:
#renaming cols 76-85

new_names=['Jump Count', 'Total Grit', 'Avg Flow', 'Flagged', 'Avg Elapsed Speed','Dirt Distance', 'Newly Explored Distance', 'Newly Explored Dirt Distance', 'Sport Type','Total Steps']

df.rename(columns=dict(zip(df.columns[76:86], new_names)), inplace=True)


In [11]:
#checking that it worked

#note we used info rather than df.columns so that we could also see the index of the columns incase some had been left out we'd know which index to refer to when renaming
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1135 entries, 0 to 1134
Data columns (total 87 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Activity ID                   1135 non-null   int64  
 1   Activity Date                 1135 non-null   object 
 2   Activity Name                 1135 non-null   object 
 3   Activity Type                 1135 non-null   object 
 4   Activity Description          459 non-null    object 
 5   Elapsed Time                  1135 non-null   int64  
 6   Distance                      1135 non-null   float64
 7   Max Heart Rate                234 non-null    float64
 8   Relative Effort               234 non-null    float64
 9   Commute                       1135 non-null   bool   
 10  Activity Private Note         21 non-null     object 
 11  Activity Gear                 60 non-null     object 
 12  Filename                      952 non-null    object 
 13  Ath

Nice! All the Columns have been renamed with out the error messages.


Now that we can read the names of all our columns, lets establish what is needed and what is not

---

# Nan-Values

First lets look for nan values.


In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1135 entries, 0 to 1134
Data columns (total 87 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Activity ID                   1135 non-null   int64  
 1   Activity Date                 1135 non-null   object 
 2   Activity Name                 1135 non-null   object 
 3   Activity Type                 1135 non-null   object 
 4   Activity Description          459 non-null    object 
 5   Elapsed Time                  1135 non-null   int64  
 6   Distance                      1135 non-null   float64
 7   Max Heart Rate                234 non-null    float64
 8   Relative Effort               234 non-null    float64
 9   Commute                       1135 non-null   bool   
 10  Activity Private Note         21 non-null     object 
 11  Activity Gear                 60 non-null     object 
 12  Filename                      952 non-null    object 
 13  Ath

We can see that many columns have "0-non-null" which makes sense as the watch likely did not even record metrics to fill these feilds. We will remove these columns.

In [13]:
#grouping all the null columns to confirm they are metrics that could not be recorded
null_columns= [col for col in df.columns if df[col].isnull().all()]

#Source code--> missing_columns = [col for col in df.columns if df[col].isnull().any()]


In [14]:
null_columns

['Bike Weight',
 'Average Positive Grade',
 'Average Negative Grade',
 'Max Heart Rate.1',
 'Max Watts',
 'Average Watts',
 'Max Temperature',
 'Total Work',
 'Uphill Time',
 'Downhill Time',
 'Other Time',
 'Type',
 'Start Time',
 'Weighted Average Power',
 'Power Count',
 'Total Weight Lifted',
 'Bike',
 'Jump Count',
 'Total Grit',
 'Avg Flow',
 'Newly Explored Distance',
 'Newly Explored Dirt Distance',
 'Sport Type']

In [15]:
#making a serires out fo our variable
pd.Series(null_columns)

0                      Bike Weight
1           Average Positive Grade
2           Average Negative Grade
3                 Max Heart Rate.1
4                        Max Watts
5                    Average Watts
6                  Max Temperature
7                       Total Work
8                      Uphill Time
9                    Downhill Time
10                      Other Time
11                            Type
12                      Start Time
13          Weighted Average Power
14                     Power Count
15             Total Weight Lifted
16                            Bike
17                      Jump Count
18                      Total Grit
19                        Avg Flow
20         Newly Explored Distance
21    Newly Explored Dirt Distance
22                      Sport Type
dtype: object

Over all we have 22 columns, most of which appear to be some sports metrics that were not able to record (i.e. `jump count`, `bike weight`, etc). We will drop these columns.

In [16]:
#dropping columns with all null values

df.drop(columns=null_columns, axis=1, inplace=True)
df.shape

(1135, 64)

We now have 64 remaining columns, and will look further to drop any other unecessary columns, and feature engineering where necessary.

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1135 entries, 0 to 1134
Data columns (total 64 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Activity ID                1135 non-null   int64  
 1   Activity Date              1135 non-null   object 
 2   Activity Name              1135 non-null   object 
 3   Activity Type              1135 non-null   object 
 4   Activity Description       459 non-null    object 
 5   Elapsed Time               1135 non-null   int64  
 6   Distance                   1135 non-null   float64
 7   Max Heart Rate             234 non-null    float64
 8   Relative Effort            234 non-null    float64
 9   Commute                    1135 non-null   bool   
 10  Activity Private Note      21 non-null     object 
 11  Activity Gear              60 non-null     object 
 12  Filename                   952 non-null    object 
 13  Athlete Weight             504 non-null    float

In [18]:
#Adjudting settings so we can see the entire set of columns/rows

pd.set_option('display.max_columns', None)

In [19]:
#code for looking into a certain column where the its not na
no_nans = df[~df['Activity Gear'].isna()]
no_nans

Activity ID              Activity Date  \
953    8055405618   Nov 1, 2022, 11:10:59 PM   
959    8065329982   Nov 4, 2022, 12:49:34 AM   
962    8074153625    Nov 5, 2022, 8:49:01 PM   
976    8102856544   Nov 11, 2022, 4:48:07 PM   
977    8114828774  Nov 13, 2022, 11:04:25 PM   
980    8124588675  Nov 16, 2022, 12:33:02 AM   
983    8134947190  Nov 17, 2022, 11:51:38 PM   
985    8147216875   Nov 21, 2022, 2:54:58 AM   
989    8159957451  Nov 23, 2022, 10:43:23 PM   
994    8172059906   Nov 26, 2022, 4:18:49 PM   
997    8182836590   Nov 29, 2022, 1:19:01 AM   
1000   8187175429  Nov 29, 2022, 11:46:16 PM   
1003   8203739641    Dec 3, 2022, 9:54:47 PM   
1010   8221815032    Dec 8, 2022, 1:55:04 AM   
1014   8237781789   Dec 11, 2022, 2:42:12 AM   
1016   8245960319  Dec 13, 2022, 11:36:40 PM   
1017   8249799115  Dec 14, 2022, 11:51:37 PM   
1019   8260340120   Dec 17, 2022, 7:34:01 PM   
1020   8265039326  Dec 18, 2022, 10:40:00 PM   
1021   8267586621   Dec 19, 2022, 5:28:34 PM   
1022   8274649423  Dec 20, 2022, 12:16:14 PM   
1024   8282646906   Dec 23, 2022, 3:49:47 PM   
1025   8290802521   Dec 25, 2022, 8:36:26 PM   
1027   8299418981   Dec 27, 2022, 7:08:30 PM   
1030   8323479675    Jan 2, 2023, 1:06:23 AM   
1031   8334649471    Jan 4, 2023, 1:49:45 AM   
1032   8348794635    Jan 5, 2023, 6:03:45 PM   
1033   8377974022   Jan 7, 2023, 11:33:26 PM   
1034   8377974819    Jan 9, 2023, 4:32:13 PM   
1036   8377976298  Jan 11, 2023, 11:11:37 PM   
1043   8414956955  Jan 18, 2023, 11:41:51 PM   
1046   8424301124   Jan 20, 2023, 8:47:31 PM   
1052   8445814872   Jan 24, 2023, 6:58:35 PM   
1055   8462396584   Jan 27, 2023, 8:11:48 PM   
1062   8499947013    Feb 3, 2023, 8:09:44 PM   
1066   8512625098   Feb 5, 2023, 10:38:02 PM   
1068   8523242014    Feb 7, 2023, 8:52:37 PM   
1071   8537947597   Feb 10, 2023, 3:49:18 PM   
1072   8551479886   Feb 12, 2023, 8:26:26 PM   
1074   8561361734   Feb 14, 2023, 6:09:04 PM   
1078   8577689234   Feb 17, 2023, 7:14:29 PM   
1083   8599473273   Feb 21, 2023, 4:45:16 PM   
1089   8615010795   Feb 24, 2023, 3:22:09 PM   
1094   8633884670  Feb 27, 2023, 11:16:46 PM   
1100   8655533326    Mar 4, 2023, 1:31:19 AM   
1103   8667134147    Mar 5, 2023, 9:38:06 PM   
1106   8677388497    Mar 7, 2023, 7:44:50 PM   
1107   8683754168    Mar 8, 2023, 7:54:30 PM   
1110   8692495588   Mar 10, 2023, 6:57:11 PM   
1113   8705604334   Mar 13, 2023, 1:10:23 AM   
1115   8714424290   Mar 14, 2023, 6:11:44 PM   
1116   8720179678   Mar 15, 2023, 6:38:25 PM   
1120   8731586168   Mar 17, 2023, 9:12:35 PM   
1123   8741813861   Mar 19, 2023, 2:27:08 PM   
1124   8742157391   Mar 19, 2023, 2:50:14 PM   
1126   8754265587   Mar 21, 2023, 6:48:53 PM   
1129   8771217929   Mar 24, 2023, 8:35:02 PM   
1131   8783182130   Mar 26, 2023, 8:30:40 PM   
1133   8788572014  Mar 27, 2023, 10:17:55 PM   
1134   8794388032   Mar 28, 2023, 9:49:46 PM   

                                          Activity Name Activity Type  \
953            CUE DREAMS AND NIGHT MARES RIGHT MF NOW😩           Run   
959                                                🌫🌫🌫🌫           Run   
962                                 Bumped into Andrew🥹           Run   
976                                       5on 1:30 walk           Run   
977                                       5 on 130 off            Run   
980                                  5x5min w 1min walk           Run   
983                nov 18th  morn (date is being weird)           Run   
985                   Fort run w out walking in a while           Run   
989                          Stepped in doggoo pies😅🥲🍫💩           Run   
994                                           Lunch Run           Run   
997                                         Evening Run           Run   
1000                              Reunion run w Zach🥹👨🏻           Run   
1003                                        GPS said ✌🏼           Run   
1010                                  GPS still trash 😅

---
Lets try to query all the columns where there are fewer than 100 not-null values

In [20]:
#creating a variable that is a count of all the non-null values
notnull_count= df.notnull().sum()

In [21]:
#creating a variable that specifies columns with a certain number of non-nulls; utilizing the .loc[] funciton on the previous variable
mostly_null= notnull_count.loc[notnull_count<100]

In [22]:
#Checking
mostly_null

Activity Private Note        21
Activity Gear                60
Number of Runs                1
Perceived Exertion            4
Perceived Relative Effort     4
Weather Observation Time     15
Weather Condition            15
Weather Temperature          15
Apparent Temperature         15
Dewpoint                     15
Humidity                     15
Weather Pressure             15
Wind Speed                   15
Wind Gust                    15
Wind Bearing                 15
Precipitation Intensity      15
Sunrise Time                 15
Sunset Time                  15
Moon Phase                   15
Precipitation Probability    15
Precipitation Type            1
Cloud Cover                  15
Weather Visibility           15
UV Index                     15
Weather Ozone                15
Total Steps                  38
Media                        46
dtype: int64

In [23]:
#dropping columns with mostly null
df.drop(columns=mostly_null.index, axis=1, inplace=True)


In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1135 entries, 0 to 1134
Data columns (total 37 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Activity ID                1135 non-null   int64  
 1   Activity Date              1135 non-null   object 
 2   Activity Name              1135 non-null   object 
 3   Activity Type              1135 non-null   object 
 4   Activity Description       459 non-null    object 
 5   Elapsed Time               1135 non-null   int64  
 6   Distance                   1135 non-null   float64
 7   Max Heart Rate             234 non-null    float64
 8   Relative Effort            234 non-null    float64
 9   Commute                    1135 non-null   bool   
 10  Filename                   952 non-null    object 
 11  Athlete Weight             504 non-null    float64
 12  Elapsed Time.1             1059 non-null   float64
 13  Moving Time                1135 non-null   float

# .1 columns

Nice, the columns were dropped and we now only have 37 columns. However upon further inspection there appear to be columns with similar names except for a **.1**, (i.e. `Commute` and `Commute.1`), we will check the similarity, and delete the potetial duplicates.

In [25]:
df[['Distance', 'Distance.1', 'Commute', 'Commute.1', 'Relative Effort', 'Relative Effort.1', 'Elapsed Time', 'Elapsed Time.1']]



Distance    Distance.1  Commute  Commute.1  Relative Effort  \
0         2.81   2818.199951    False        0.0              NaN   
1         2.14   2148.600098    False        0.0              NaN   
2        13.72  13723.299805    False        0.0              NaN   
3         8.82   8828.099609    False        0.0              NaN   
4         7.85   7850.200195    False        0.0              NaN   
...        ...           ...      ...        ...              ...   
1130      0.00      0.000000    False        0.0             26.0   
1131      7.60   7599.910156    False        0.0             56.0   
1132      0.00      0.000000    False        0.0             39.0   
1133      3.62   3620.290039    False        0.0             16.0   
1134      8.12   8122.649902    False        0.0             62.0   

      Relative Effort.1  Elapsed Time  Elapsed Time.1  
0                   NaN          1310          1310.0  
1                   NaN           663           663.0  
2                   NaN          4658          4658.0  
3                   NaN          2924          2924.0  
4                   NaN          2305          2305.0  
...                 ...           ...             ...  
1130               26.0          3526          3526.0  
1131               56.0          2259          2259.0  
1132               39.0          3790          3790.0  
1133               16.0          1211          1211.0  
1134               62.0          2336          2336.0  

[1135 rows x 8 columns]

It appears that the columns have identical data apart from the decimal place; except for commute which either labels it as a T/F or 1/0. For the sake of simplicity we will remove the '.1' columns as they provide the same information as their counter parts apart from granularity.

In [26]:
#droppping repeat .1 columns
df = df[df.columns.drop(list(df.filter(regex='.1')))]


In [27]:
df.shape

(1135, 33)

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1135 entries, 0 to 1134
Data columns (total 33 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Activity ID                1135 non-null   int64  
 1   Activity Date              1135 non-null   object 
 2   Activity Name              1135 non-null   object 
 3   Activity Type              1135 non-null   object 
 4   Activity Description       459 non-null    object 
 5   Elapsed Time               1135 non-null   int64  
 6   Distance                   1135 non-null   float64
 7   Max Heart Rate             234 non-null    float64
 8   Relative Effort            234 non-null    float64
 9   Commute                    1135 non-null   bool   
 10  Filename                   952 non-null    object 
 11  Athlete Weight             504 non-null    float64
 12  Moving Time                1135 non-null   float64
 13  Max Speed                  952 non-null    float

---
Now we are down to 32 columns. I still feel as if this is too many columns, I will take a look at the data frame again and try to pick out columns that are unnecessary based on my knowledge of what my watch records, and what info we are interested in. Additionally I will put a focus on removing columns that do not have 1135 rows, meaning they likely have many null values.

In [29]:
#making a varaible for any 
notnull_count= df.notnull().sum()

In [30]:
# Making a variable for columns with any null values
any_null= notnull_count.loc[notnull_count<1135]

In [31]:
#sorting columns with null values
any_null.sort_values()

Max Heart Rate               234
Relative Effort              234
Average Heart Rate           234
Average Temperature          287
Flagged                      380
Avg Elapsed Speed            390
Prefer Perceived Exertion    433
Activity Description         459
Grade Adjusted Distance      480
Athlete Weight               504
From Upload                  650
Average Speed                679
Gear                         680
Dirt Distance                698
Max Cadence                  804
Average Cadence              804
Elevation High               898
Elevation Low                898
Elevation Loss               898
Elevation Gain               928
Max Speed                    952
Filename                     952
Max Grade                    952
Calories                     992
dtype: int64

Off the bat, I want to remove `max heart rate` (as i doubt the accuracy of my hr monitor), `Calories` (as that is an estimation and a crude one from what I understand), `filename`, `From Upload`, and more but I want to look further to understand the columns a little better before dropping them.

---
Lets look into the numbers based on activity type to see if anyhting lines up. 

In [32]:
df['Activity Type'].value_counts()

Run                708
Ride               194
Workout            113
Weight Training     76
Elliptical          23
Hike                11
Nordic Ski           7
Virtual Ride         1
Snowboard            1
Ice Skate            1
Name: Activity Type, dtype: int64

Just visually I dont see any of these numbers lining up with the numbers in the any_null group (i.e. if max speed only had 170 non-nulls, and cycling had 170 values then i'd know they were associated and my watch may not record that value for that activity). 

**Note** :Is there a better way I could have looked for these numbers?
I guess I could have looped through a list of activity type counts and compared it to null counts.


In [33]:
#ask senpai alex

'''activity_type_count=df['Activity Type'].value_counts()

for x in activity_type_count.values():
    if x==mostly_null.values():
        print ('match found')
'''

#do i use enumerate? or convert to a dctionary? 
# im jsut tryna match the values in one to the values in another
#if x ==

"activity_type_count=df['Activity Type'].value_counts()\n\nfor x in activity_type_count.values():\n    if x==mostly_null.values():\n        print ('match found')\n"

After some technical difficulties I have decided to stick with my initial method of comparison.

# Moving on!

There appear to be no matches, so we can not attriubte any of the nulls in a column ot jsut one activity type.

From the list, the only columns I deem worth keeping are,`Activty Description`, `Max Grade`, `Max Speed`, `Elevation Gain`, `Elevation Loss`, `Average Cadence`, `Max Heart Rate`, `Average Heart Rate` and `Gear` (although this is so low i figure it could be intersting to see what shoes were worn during the course of injury). The rest either seem either too infreuwently reported, werent measured for, or seem irrelevant.

In [34]:
any_null.sort_values()

Max Heart Rate               234
Relative Effort              234
Average Heart Rate           234
Average Temperature          287
Flagged                      380
Avg Elapsed Speed            390
Prefer Perceived Exertion    433
Activity Description         459
Grade Adjusted Distance      480
Athlete Weight               504
From Upload                  650
Average Speed                679
Gear                         680
Dirt Distance                698
Max Cadence                  804
Average Cadence              804
Elevation High               898
Elevation Low                898
Elevation Loss               898
Elevation Gain               928
Max Speed                    952
Filename                     952
Max Grade                    952
Calories                     992
dtype: int64

In [35]:
#getting a list of mostly_null
any_null_list=list(any_null.index)

In [36]:
any_null_list

['Activity Description',
 'Max Heart Rate',
 'Relative Effort',
 'Filename',
 'Athlete Weight',
 'Max Speed',
 'Average Speed',
 'Elevation Gain',
 'Elevation Loss',
 'Elevation Low',
 'Elevation High',
 'Max Grade',
 'Max Cadence',
 'Average Cadence',
 'Average Heart Rate',
 'Calories',
 'Average Temperature',
 'Prefer Perceived Exertion',
 'From Upload',
 'Grade Adjusted Distance',
 'Gear',
 'Flagged',
 'Avg Elapsed Speed',
 'Dirt Distance']

In [37]:
len(any_null_list)

24

In [38]:
keepers=['Max Grade','Max Speed', 'Elevation Gain', 'Elevation Loss', 'Average Cadence', 'Average Speed', 'Activity Description', 'Max Heart Rate', 'Average Heart Rate']


In [39]:
len(keepers)

9

So should get 15 back

In [40]:
for i in keepers:
    any_null_list.remove(i)

print(len(any_null_list))

15


In [41]:
any_null_list

['Relative Effort',
 'Filename',
 'Athlete Weight',
 'Elevation Low',
 'Elevation High',
 'Max Cadence',
 'Calories',
 'Average Temperature',
 'Prefer Perceived Exertion',
 'From Upload',
 'Grade Adjusted Distance',
 'Gear',
 'Flagged',
 'Avg Elapsed Speed',
 'Dirt Distance']

In [42]:
len(any_null_list)

15

Nice, those columns have beeen dropped from our list, which we will use to drop the columns from the dataframe.

In [43]:
#checkign dataframe shape
df.shape

(1135, 33)

In [44]:
df.shape[1]- len(any_null_list)

18

Should be left with 18 columns.

In [45]:
#dropping the columns
df.drop(columns=any_null_list, axis=1, inplace=True)
df.shape

/var/folders/33/fr98d58s73v34qbn6dcdzx380000gn/T/ipykernel_19984/3999182295.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(columns=any_null_list, axis=1, inplace=True)


(1135, 18)

Nice, lets check

In [46]:
df.head()

Activity ID              Activity Date       Activity Name Activity Type  \
0   2081785707   Nov 17, 2018, 5:04:51 PM      Lunch Activity       Workout   
1   2081785716  Dec 27, 2018, 10:36:08 PM  Afternoon Activity       Workout   
2   2081785724    Nov 7, 2018, 9:02:38 PM  Afternoon Activity       Workout   
3   2081785725   Dec 18, 2018, 9:45:51 PM  Afternoon Activity       Workout   
4   2081785728    Jan 6, 2019, 8:19:24 PM  Afternoon Activity       Workout   

  Activity Description  Elapsed Time  Distance  Max Heart Rate  Commute  \
0                  NaN          1310      2.81             NaN    False   
1                  NaN           663      2.14             NaN    False   
2                  NaN          4658     13.72             NaN    False   
3                  NaN          2924      8.82             NaN    False   
4                  NaN          2305      7.85             NaN    False   

   Moving Time  Max Speed  Average Speed  Elevation Gain  Elevation Loss  \
0        922.0        8.6            NaN            14.0            25.0   
1        663.0        3.8            NaN             2.0             0.0   
2       4244.0        4.0            NaN           149.0           156.0   
3       2685.0        4.4            NaN            41.0            44.0   
4       2195.0        4.4            NaN             6.0            13.0   

   Max Grade  Average Grade  Average Cadence  Average Heart Rate  
0  27.299999      -0.383223        80.825447                 NaN  
1  18.700001       0.121009        81.068810                 NaN  
2  15.300000      -0.049551        80.853722                 NaN  
3  17.700001      -0.011327        80.129654                 NaN  
4   4.700000      -0.078979        82.909218                 NaN

Last, we will also drop activity_ID. While it has no nulls, it provides no extra information that can not be obtained form the other columns or the index itself.

In [47]:
df.drop(columns='Activity ID', inplace=True)

/var/folders/33/fr98d58s73v34qbn6dcdzx380000gn/T/ipykernel_19984/733528720.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(columns='Activity ID', inplace=True)


In [48]:
df

Activity Date                               Activity Name  \
0      Nov 17, 2018, 5:04:51 PM                              Lunch Activity   
1     Dec 27, 2018, 10:36:08 PM                          Afternoon Activity   
2       Nov 7, 2018, 9:02:38 PM                          Afternoon Activity   
3      Dec 18, 2018, 9:45:51 PM                          Afternoon Activity   
4       Jan 6, 2019, 8:19:24 PM                          Afternoon Activity   
...                         ...                                         ...   
1130   Mar 25, 2023, 2:26:39 PM  Album: Sticky fingers- Land of pleasure \n   
1131   Mar 26, 2023, 8:30:40 PM                              Takin it easy    
1132   Mar 27, 2023, 3:33:02 PM                                     Cyclone   
1133  Mar 27, 2023, 10:17:55 PM                     Good to see tha homies    
1134   Mar 28, 2023, 9:49:46 PM             Zach gate has been proven wrong   

     Activity Type                               Activity Description  \
0          Workout                                                NaN   
1          Workout                                                NaN   
2          Workout                                                NaN   
3          Workout                                                NaN   
4          Workout                                                NaN   
...            ...                                                ...   
1130          Ride             15 wu\n10-8-6 w 2\n15 cd\n\n90-100 rpm   
1131           Run  Didn’t go as far, hoping the calf feels a bit ...   
1132    Elliptical                  15 wu\n8 tempo\n3\n8x2 w 1\n15 cd   
1133           Run                                                NaN   
1134           Run                                                NaN   

      Elapsed Time  Distance  Max Heart Rate  Commute  Moving Time  Max Speed  \
0             1310      2.81             NaN    False        922.0   8.600000   
1              663      2.14             NaN    False        663.0   3.800000   
2             4658     13.72             NaN    False       4244.0   4.000000   
3             2924      8.82             NaN    False       2685.0   4.400000   
4             2305      7.85             NaN    False       2195.0   4.400000   
...            ...       ...             ...      ...          ...        ...   
1130          3526      0.00           154.0    False       3526.0   0.000000   
1131          2259      7.60           174.0    False       2133.0   7.054004   
1132          3790      0.00           167.0    False       3790.0   0.000000   
1133          1211      3.62           161.0    False       1195.0   4.691671   
1134          2336      8.12           184.0    False       2234.0   7.006055   

      Average Speed  Elevation Gain  Elevation Loss  Max Grade  Average Grade  \
0               NaN       14.000000            25.0  27.299999      -0.383223   
1               NaN        2.000000             0.0  18.700001       0.121009   
2               NaN      149.000000           156.0  15.300000      -0.049551   
3               NaN       41.000000            44.0  17.700001      -0.011327   
4               NaN        6.000000            13.0   4.700000      -0.078979   
...             ...             ...             ...        ...            ...   
1130       0.000000        0.000000             NaN   0.000000       0.000000   
1131       3.563015       34.896828           589.0   4.860952       0.021053   
1132       0.000000        0.000000             NaN   0.000000       0.000000   
1133       3.029531      144.000000           188.0  29.058149      -1.099360   
1134       3.635922       35.247936           403.0   5.430344      -0.033240   

      Average Cadence  Average Heart Rate  
0           80.825447                 NaN  
1           81.068810                 NaN  
2           80.853722                 NaN  
3           80.129654                 NaN  
4           82.909218     

---

Note: Although there are some columns like elapsed time are in seconds and Distance is in km. I debate either noting this in the column header as well as converting the measurement in seconds to minutes. 

## Changing the time to Minutes
We can always change it back to seconds if needed, and this makes the values readily interpretable.

In [49]:
#dividing the elapsed time col by 60 so that it can be read as minutes
df['Elapsed Time']=df['Elapsed Time']/60

/var/folders/33/fr98d58s73v34qbn6dcdzx380000gn/T/ipykernel_19984/2846644504.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Elapsed Time']=df['Elapsed Time']/60


In [50]:
#checking 
df.tail(2)

Activity Date                    Activity Name  \
1133  Mar 27, 2023, 10:17:55 PM          Good to see tha homies    
1134   Mar 28, 2023, 9:49:46 PM  Zach gate has been proven wrong   

     Activity Type Activity Description  Elapsed Time  Distance  \
1133           Run                  NaN     20.183333      3.62   
1134           Run                  NaN     38.933333      8.12   

      Max Heart Rate  Commute  Moving Time  Max Speed  Average Speed  \
1133           161.0    False       1195.0   4.691671       3.029531   
1134           184.0    False       2234.0   7.006055       3.635922   

      Elevation Gain  Elevation Loss  Max Grade  Average Grade  \
1133      144.000000           188.0  29.058149       -1.09936   
1134       35.247936           403.0   5.430344       -0.03324   

      Average Cadence  Average Heart Rate  
1133        79.307625          140.206055  
1134        81.354897          154.254822

Nice, we can see our last run was 38.9 minutes. We will do the same to the other time column `Moving Time`.

In [51]:
#converting the Moving Time as well

df['Moving Time']=df['Moving Time']/60

/var/folders/33/fr98d58s73v34qbn6dcdzx380000gn/T/ipykernel_19984/3901295577.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Moving Time']=df['Moving Time']/60


Decided to leave the distance header alone.

---
Now we will export this copy of the data frame to be worked with later.

In [52]:
# exporting a copy of this data to be used in our models

#commented this out so that we cna re run notebook to check things w out exporting more copies of the same dataset
'''
df.to_csv('Reduced_strava_df_V2.csv', index=None)
'''

"\ndf.to_csv('Reduced_strava_df_V2.csv', index=None)\n"

In [53]:
df.tail(10)

Activity Date  \
1125  Mar 21, 2023, 12:20:44 AM   
1126   Mar 21, 2023, 6:48:53 PM   
1127   Mar 22, 2023, 9:33:30 PM   
1128  Mar 24, 2023, 12:05:27 AM   
1129   Mar 24, 2023, 8:35:02 PM   
1130   Mar 25, 2023, 2:26:39 PM   
1131   Mar 26, 2023, 8:30:40 PM   
1132   Mar 27, 2023, 3:33:02 PM   
1133  Mar 27, 2023, 10:17:55 PM   
1134   Mar 28, 2023, 9:49:46 PM   

                                          Activity Name Activity Type  \
1125                                       Evening Ride          Ride   
1126                                  I gotta roll more           Run   
1127  Calf felt a bit janky, opted for the bike instead          Ride   
1128                 Mural is an OP cross training song    Elliptical   
1129                                   Tight gastroccys           Run   
1130         Album: Sticky fingers- Land of pleasure \n          Ride   
1131                                     Takin it easy            Run   
1132                                            Cyclone    Elliptical   
1133                            Good to see tha homies            Run   
1134                    Zach gate has been proven wrong           Run   

                                   Activity Description  Elapsed Time  \
1125                2x(4-3-2) w 1:30 and 3 between sets     57.266667   
1126                                                NaN     57.233333   
1127                                                NaN     60.250000   
1128  Good tempo, fat bars, hella long. What more ca...     60.516667   
1129                                                NaN     38.650000   
1130             15 wu\n10-8-6 w 2\n15 cd\n\n90-100 rpm     58.766667   
1131  Didn’t go as far, hoping the calf feels a bit ...     37.650000   
1132                  15 wu\n8 tempo\n3\n8x2 w 1\n15 cd     63.166667   
1133                                                NaN     20.183333   
1134                                                NaN     38.933333   

      Distance  Max Heart Rate  Commute  Moving Time  Max Speed  \
1125      0.00           166.0    False    57.266667   0.000000   
1126     10.00           170.0    False    46.650000  10.087891   
1127      0.00           145.0    False    60.250000   0.000000   
1128      0.00           167.0    False    60.516667   0.000000   
1129      8.15           168.0    False    37.216667   7.706055   
1130      0.00           154.0    False    58.766667   0.000000   
1131      7.60           174.0    False    35.550000   7.054004   
1132      0.00           167.0    False    63.166667   0.000000   
1133      3.62           161.0    False    19.916667   4.691671   
1134      8.12           184.0    False    37.233333   7.006055   

      Average Speed  Elevation Gain  Elevation Loss  Max Grade  Average Grade  \
1125       0.000000        0.000000             NaN   0.000000       0.000000   
1126       3.574845      598.000000           867.0  49.739868      -0.161903   
1127       0.000000        0.000000             NaN   0.000000       0.000000   
1128       0.000000        0.000000             NaN   0.000000       0.000000   
1129       3.651554       37.175983           588.0   5.926308      -0.012264   
1130       0.000000        0.000000             NaN   0.000000       0.000000   
1131       3.563015       34.896828           589.0   4.860952       0.021053   
1132       0.000000        0.000000             NaN   0.000000       0.000000   
1133       3.029531      144.000000           188.0  29.058149      -1.099360   
1134       3.635922       35.247936           403.0   5.430344      -0.033240   

      Average Cadence  Average Heart Rate  
1125              NaN          126.123398  
1126        79.508858          148.332855  
1127              NaN          119.373169  
1128        77.363388          135.979889  
1129        79.577858          154.194962  
1130              NaN          122.743904  
1131        79.739670          154.208176  
1132        75.642998          132.537994 

---
## Looking back at 2019 to understand why there ia no run speciffic data 


In [54]:
#converting date column to date time
df['Activity Date']=pd.to_datetime(df['Activity Date'])

/var/folders/33/fr98d58s73v34qbn6dcdzx380000gn/T/ipykernel_19984/2314097987.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Activity Date']=pd.to_datetime(df['Activity Date'])


In [55]:
#checking
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1135 entries, 0 to 1134
Data columns (total 17 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Activity Date         1135 non-null   datetime64[ns]
 1   Activity Name         1135 non-null   object        
 2   Activity Type         1135 non-null   object        
 3   Activity Description  459 non-null    object        
 4   Elapsed Time          1135 non-null   float64       
 5   Distance              1135 non-null   float64       
 6   Max Heart Rate        234 non-null    float64       
 7   Commute               1135 non-null   bool          
 8   Moving Time           1135 non-null   float64       
 9   Max Speed             952 non-null    float64       
 10  Average Speed         679 non-null    float64       
 11  Elevation Gain        928 non-null    float64       
 12  Elevation Loss        898 non-null    float64       
 13  Max Grade         

In [56]:
df[(df['Activity Date'].dt.year==2019) & (df['Activity Type']== 'Run')]
#this didnt work

Activity Date        Activity Name Activity Type  \
109 2019-02-18 17:09:10              Warm up           Run   
110 2019-02-18 18:31:22            Cool down           Run   
111 2019-02-19 16:48:52  Double cemetery run           Run   
112 2019-02-20 22:55:10   Afternoon Activity           Run   
113 2019-02-21 00:28:13     Evening Activity           Run   
..                  ...                  ...           ...   
515 2019-12-30 22:24:53        Afternoon Run           Run   
516 2019-12-30 23:05:30          Evening Run           Run   
517 2019-12-30 23:46:27          Evening Run           Run   
518 2019-12-31 18:22:06        Afternoon Run           Run   
523 2019-12-31 20:00:31            Treadmill           Run   

    Activity Description  Elapsed Time  Distance  Max Heart Rate  Commute  \
109                  NaN     20.783333      3.01             NaN    False   
110                  NaN     18.566667      3.34             NaN    False   
111                  NaN     80.800000     15.00             NaN    False   
112                  NaN     10.433333      2.06             NaN    False   
113                  NaN     17.533333      3.32             NaN    False   
..                   ...           ...       ...             ...      ...   
515                  NaN     13.700000      2.05             NaN    False   
516                  NaN     32.366667      6.02             NaN    False   
517                  NaN     16.500000      2.16             NaN    False   
518                  NaN     88.583333     15.51             NaN    False   
523                  NaN     25.000000      5.00             NaN    False   

     Moving Time  Max Speed  Average Speed  Elevation Gain  Elevation Loss  \
109    17.050000        3.3            NaN      155.399994            11.0   
110    17.666667        3.9            NaN       14.400000             7.0   
111    71.283333        5.2            NaN      167.800003           120.0   
112    10.433333        3.9            NaN        4.000000             1.0   
113    17.533333        4.5            NaN        9.600000             5.0   
..           ...        ...            ...             ...             ...   
515    11.716667        6.2       2.503000        0.000000           104.0   
516    27.666667        6.5       3.104000        0.000000           177.0   
517    13.066667        4.6       2.758000        0.000000            59.0   
518    72.150000       10.2       3.584000       73.196663           431.0   
523    25.000000        NaN       3.333333             NaN             NaN   

     Max Grade  Average Grade  Average Cadence  Average Heart Rate  
109   4.600000      -0.079576        82.033562                 NaN  
110   6.800000      -0.077802        82.502853                 NaN  
111  22.200001       0.047999        83.357735                 NaN  
112  11.400000       0.135797        80.692055                 NaN  
113   4.900000       0.030117        80.510071                 NaN  
..         ...            ...              ...                 ...  
515   1.900000      -0.004861        81.123756                 NaN  
516   2.900000       0.014932        86.266495                 NaN  
517   1.800000       0.000000        80.803574                 NaN  
518   5.400000       0.000000        82.999535                 NaN  
523        NaN       0.000000              NaN                 NaN  

[388 rows x 17 columns]

---
Hmm as suspected my old watch did not record average speed. We could fill this value with a simple calculation between the distance ran and the time ((Distance*1000)/(Moving Time*60) which would give us speed in m/s which is how it is recorded for the other columns. Lets go for it.

In [59]:
#checking number of NANs in average speed atm
df['Average Speed'].isna().sum()

456

There are 456 NAN values to be replaced.

In [60]:
#create a variable for our average speed calculation
avg_speed_calc = (df['Distance']*1000)/(df['Moving Time']*60)

#using .loc to fillna values with our speed calculation
df.loc[:, "Average Speed"] = df.loc[:, "Average Speed"].fillna(avg_speed_calc)

/var/folders/33/fr98d58s73v34qbn6dcdzx380000gn/T/ipykernel_19984/4155283065.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, "Average Speed"] = df.loc[:, "Average Speed"].fillna(avg_speed_calc)


In [61]:
#checking number of NAN's left over
df['Average Speed'].isna().sum()

0

NONE! Nice!

In [63]:
#rechcking our 2019 columns that had issues
df[(df['Activity Date'].dt.year==2019) & (df['Activity Type']== 'Run')]


Activity Date        Activity Name Activity Type  \
109 2019-02-18 17:09:10              Warm up           Run   
110 2019-02-18 18:31:22            Cool down           Run   
111 2019-02-19 16:48:52  Double cemetery run           Run   
112 2019-02-20 22:55:10   Afternoon Activity           Run   
113 2019-02-21 00:28:13     Evening Activity           Run   
..                  ...                  ...           ...   
515 2019-12-30 22:24:53        Afternoon Run           Run   
516 2019-12-30 23:05:30          Evening Run           Run   
517 2019-12-30 23:46:27          Evening Run           Run   
518 2019-12-31 18:22:06        Afternoon Run           Run   
523 2019-12-31 20:00:31            Treadmill           Run   

    Activity Description  Elapsed Time  Distance  Max Heart Rate  Commute  \
109                  NaN     20.783333      3.01             NaN    False   
110                  NaN     18.566667      3.34             NaN    False   
111                  NaN     80.800000     15.00             NaN    False   
112                  NaN     10.433333      2.06             NaN    False   
113                  NaN     17.533333      3.32             NaN    False   
..                   ...           ...       ...             ...      ...   
515                  NaN     13.700000      2.05             NaN    False   
516                  NaN     32.366667      6.02             NaN    False   
517                  NaN     16.500000      2.16             NaN    False   
518                  NaN     88.583333     15.51             NaN    False   
523                  NaN     25.000000      5.00             NaN    False   

     Moving Time  Max Speed  Average Speed  Elevation Gain  Elevation Loss  \
109    17.050000        3.3       2.942326      155.399994            11.0   
110    17.666667        3.9       3.150943       14.400000             7.0   
111    71.283333        5.2       3.507131      167.800003           120.0   
112    10.433333        3.9       3.290735        4.000000             1.0   
113    17.533333        4.5       3.155894        9.600000             5.0   
..           ...        ...            ...             ...             ...   
515    11.716667        6.2       2.503000        0.000000           104.0   
516    27.666667        6.5       3.104000        0.000000           177.0   
517    13.066667        4.6       2.758000        0.000000            59.0   
518    72.150000       10.2       3.584000       73.196663           431.0   
523    25.000000        NaN       3.333333             NaN             NaN   

     Max Grade  Average Grade  Average Cadence  Average Heart Rate  
109   4.600000      -0.079576        82.033562                 NaN  
110   6.800000      -0.077802        82.502853                 NaN  
111  22.200001       0.047999        83.357735                 NaN  
112  11.400000       0.135797        80.692055                 NaN  
113   4.900000       0.030117        80.510071                 NaN  
..         ...            ...              ...                 ...  
515   1.900000      -0.004861        81.123756                 NaN  
516   2.900000       0.014932        86.266495                 NaN  
517   1.800000       0.000000        80.803574                 NaN  
518   5.400000       0.000000        82.999535                 NaN  
523        NaN       0.000000              NaN                 NaN  

[388 rows x 17 columns]

In [65]:
#lets check the tail to make sure recent dates were not affected
df.tail(4)

Activity Date                    Activity Name Activity Type  \
1131 2023-03-26 20:30:40                   Takin it easy            Run   
1132 2023-03-27 15:33:02                          Cyclone    Elliptical   
1133 2023-03-27 22:17:55          Good to see tha homies            Run   
1134 2023-03-28 21:49:46  Zach gate has been proven wrong           Run   

                                   Activity Description  Elapsed Time  \
1131  Didn’t go as far, hoping the calf feels a bit ...     37.650000   
1132                  15 wu\n8 tempo\n3\n8x2 w 1\n15 cd     63.166667   
1133                                                NaN     20.183333   
1134                                                NaN     38.933333   

      Distance  Max Heart Rate  Commute  Moving Time  Max Speed  \
1131      7.60           174.0    False    35.550000   7.054004   
1132      0.00           167.0    False    63.166667   0.000000   
1133      3.62           161.0    False    19.916667   4.691671   
1134      8.12           184.0    False    37.233333   7.006055   

      Average Speed  Elevation Gain  Elevation Loss  Max Grade  Average Grade  \
1131       3.563015       34.896828           589.0   4.860952       0.021053   
1132       0.000000        0.000000             NaN   0.000000       0.000000   
1133       3.029531      144.000000           188.0  29.058149      -1.099360   
1134       3.635922       35.247936           403.0   5.430344      -0.033240   

      Average Cadence  Average Heart Rate  
1131        79.739670          154.208176  
1132        75.642998          132.537994  
1133        79.307625          140.206055  
1134        81.354897          154.254822

Nice! Now that we've gotten our average speed issues fixed, lets export this DF (again) and use it in tableau.



In [66]:
#Exporting refined csv of data
df.to_csv('Reduced_strava_df_V3.csv', index=None)


---

Conclusion

Defintely lots of routes I could take to further expand on this, this but seeing as it was one of my first projects back and the data set was relatively small I figured I'd stick to basic eda and visualizaitons on tableau.

If I find more time in the future it could be coolt to look into the Activity Descriptions and see which words are most common (i.e. 'good', 'calf', 'achilles', etc.).
